File: app/summarization.py
from sumy.parsers.plaintext import PlaintextParser
from sumy.nlp.tokenizers import Tokenizer
from sumy.summarizers.lsa import LsaSummarizer
from transformers import pipeline

def extractive_summary(text, sentence_count=5):
    parser = PlaintextParser.from_string(text, Tokenizer('english'))
    summarizer = LsaSummarizer()
    summary = summarizer(parser.document, sentence_count)
    return ' '.join([str(sentence) for sentence in summary])

summarizer = pipeline('summarization', model='facebook/bart-large-cnn')

def abstractive_summary(text):
    summary = summarizer(text, max_length=150, min_length=30, do_sample=False)
    return summary[0]['summary_text']

File: js/app/summarization.js
const { SummarizerManager } = require("node-summarizer");
const { pipeline } = require("@xenova/transformers");

async function extractiveSummary(text) {
  const summarizer = new SummarizerManager(text, 5);
  const summary = await summarizer.getSummaryByRank();
  const nounsAndAdjectiveMap = {};

  summary.nouns_and_adjactive_map.forEach((value, key) => {
    nounsAndAdjectiveMap[key] = value;
  });

  return {
    summary: summary.summary,
    sentence_list: summary.sentence_list,
    nouns_and_adjactive_map: nounsAndAdjectiveMap,
  };
}

async function abstractiveSummary(text) {
  const summarizer = await pipeline("summarization", "facebook/bart-large-cnn");
  const summary = await summarizer(text, { max_length: 150, min_length: 30 });

  return summary;
}

module.exports = { extractiveSummary, abstractiveSummary };
